lib/filter/filter-pipe.c

```diff
@@ -90,7 +90,18 @@ static LogPipe *
 log_filter_pipe_clone(LogPipe *s)
 {
   LogFilterPipe *self = (LogFilterPipe *) s;
-  LogPipe *cloned = log_filter_pipe_new(filter_expr_ref(self->expr), s->cfg);
+  FilterExprNode *expr;
+
+  if (self->expr->clone)
```

Comment:
I'm sorry to be late for the party, but there are some hidden traps near the filters. The filters can be referenced in many places (see `filter_expr_ref`). In addition the filters has their own sub-language in syslog-ng grammar. For example a filter could just reference another filter and combined multiple filters with logical operators.

A strange configuration example:
```
@version: 3.30

filter rl {
  throttle(
    rate(1)
    value(MESSAGE)
  );
};

filter rl_ref {
  filter(rl);
};

filter rl_or {
  filter(rl) or filter(rl);
};

log {
source { example-msg-generator( freq(0.001) ); };

rewrite {
   set("please drop me" value("MESSAGE") condition( filter(rl_ref) )); # not cloned
};
filter(rl_ref); # not cloned
filter(rl_or); # not cloned
filter(rl); # correctly cloned

destination { file("/dev/stdout"); };
};
```

As the `FilterExprNode` can be combined with `and`, `ref`, `or` and `not` kinda operators, it is a dag. As a result, you could only use a reference of a `FilterExprNode` if any of its sub expression can be used as a reference (no clone implemented), otherwise you have to do a clone.

Reply:
Looking at this now - apologies for the delay. TIL that filters are a lot more complicated than I thought

Reply:
From what I understand (and bearing in mind I'm just now starting to understand the complexity of filter composition), would an appropriate solution here be to change `filter_expr_ref` to be:

```c
FilterExprNode *
filter_expr_ref(FilterExprNode *self)
{
  if (self->clone)
    {
      return self->clone(self);
    }

  self->ref_cnt++;
  return self;
}
```

i.e. disabling multiple references for clonable `FilterExprNode`s?

Reply:
Hi again, @sinkingpoint,

We've discussed this matter IRL with @bazsi and @alltilla, and we agreed that cloning the filter object would be the best possible action here, especially because repeating the filter or just using boolean algebra on filters with state  (`filter(rl) or filter(rl)`) would cause difficult-to-predict behavior otherwise.

If `filter_expr_ref()` cloned the object conditionally, it would be a bit suspicious for me; but I think the same could work the other way around:

We could probably add and use `filter_expr_clone()` instead of `filter_expr_ref()` when searching for a named filter inside `filter_call_init()`.
`filter_expr_clone()` could safely fall back to `filter_expr_ref()` if the clone virtual method is not implemented.

@bazsi @alltilla @Kokan  What do you think?

Reply:
Sorry, I forgot to mention the most important part: we should implement `clone()` for all boolean operators in a recursive manner, so they either clone or reference their left and right operands.

Reply:
I agree with adding `clone`, in fact I would add `clone` and simple remove the `ref` interfaces. (always doing clone)
I would really like to reduce complexity and not increasing it.
There were concerns regarding boilerplate, and possible shared state filters in the future. While the boilerplate would indeed placed, but very close to the code it is written, so it would not increase the complexity. Sharing state would still be possible between cloned filters if needed.